### PR TITLE
www/caddy: Layer4 support in listener_wrapper to route TLS traffic without termination

### PR DESCRIPTION
**Pull request by Monviech**
Reference: https://github.com/mholt/caddy-l4/commit/ec5408af7e7e2458c6e1af586f7d679573e5cf3a

Requires: https://github.com/opnsense/tools/pull/423

Layer 4 support has come to the Caddyfile.

This PR implements these matchers:

```
layer4.matchers.http - matches connections that start with HTTP requests. In addition, any [http.matchers modules](https://caddyserver.com/docs/modules/) can be used for matching on HTTP-specific properties of requests, such as header or path. Note that only the first request of each connection can be used for matching.

layer4.matchers.tls - matches connections that start with TLS handshakes. In addition, any [tls.handshake_match modules](https://caddyserver.com/docs/modules/) can be used for matching on TLS-specific properties of the ClientHello, such as ServerName (SNI).
```

This commit implements the most simple form of Layer 4 that makes the most sense for this plugin: 
**_SNI based matching of TCP/UDP traffic that uses the TLS protocol_**

**The routing works like this:**

1. `Layer4 App` matches all traffic, inspects the Client Hello (TLS SNI)
2. Layer4 routes can pass this traffic unaltered to upstreams (No TLS termination of traffic)
3. A default route passes all other traffic to the `HTTP App`
4. Reverse Proxy handles HTTP traffic for all other domains normally

**This implementation stays simple for the biggest usecase:**
- Proxy with TLS termination (HTTP App)
- Proxy without TLS termination (Layer4 App)

To keep the plugin concise, Layer4 support has to be enabled in the advanced options of General Settings. Afterwards, the Layer4 Routes tab will appear. This will hide the option from first time users (hopefully). That also means, the compiled extra module can be removed at any time if it makes trouble, without affecting the main Caddy application.

- [x] **Implement and test `tls sni` matcher**
- **Usecase:** Some specific domains should be handled by upstreams directly without tls termination, all other domains should be handled by the HTTP reverse proxy.

```
{
	servers {
		listener_wrappers {
			layer4 {
				import /usr/local/etc/caddy/caddy.d/*.layer4
				@f89842cb-5bfb-4185-a24a-c1164dc316f1-tls tls sni opnsense02.example.com
				route @f89842cb-5bfb-4185-a24a-c1164dc316f1-tls {
					proxy 172.16.0.104:8080 {
					}
				}
				route
			}
			tls
		}
	}
}

# Reverse Proxy Domain: "e48549d8-ed37-4f68-8db3-ec7c771557ca"
opnsense01.example.com {
	handle {
		reverse_proxy 172.16.0.104:8080 {
			transport http {
				tls_insecure_skip_verify
			}
		}
	}
}
```
**This configuration explained:**
- 1. sends all TLS traffic directed to `opnsense02.example.com` unaltered to `172.16.0.104:8080` (without TLS termination). Browser will show internal certificate of OPNsense.
- 2. sends all HTTP traffic directed to `opnsense01.example.com` to `172.16.0.104:8080` (with TLS termination). Browser will show Let's Encrypt certificate.

- [x] **Implement and test `not tls sni` matcher**
- **Usecase:** Some specific domains shall be handled by the HTTP reverse proxy, all other domains should be passed to a hosting panel (e.g.) without TLS termination

```
{
	servers {
		listener_wrappers {
			layer4 {
				import /usr/local/etc/caddy/caddy.d/*.layer4
				@f89842cb-5bfb-4185-a24a-c1164dc316f1-tls not tls sni *.example.com *.opnsense.com
				route @f89842cb-5bfb-4185-a24a-c1164dc316f1-tls {
					proxy 172.16.0.104:8080 {
					}
				}
				route
			}
			tls
		}
	}
}

# Reverse Proxy Domain: "e48549d8-ed37-4f68-8db3-ec7c771557ca"
opnsense01.example.com {
	handle {
		reverse_proxy 172.16.0.104:8080 {
			transport http {
				tls_insecure_skip_verify
			}
		}
	}
}
```
**This configuration explained:**
- 1. sends all TLS traffic NOT directed to `*.example.com` or `*.opnsense.com` unaltered to `172.16.0.104:8080` (without TLS termination).
- 2. sends all HTTP traffic directed to `opnsense01.example.com` to `172.16.0.104:8080` (with TLS termination). Browser will show Let's Encrypt certificate.

- [x] **Implement HostnameField instead of ModelRelation Field.** The routes have to allow multiple domains per matcher. So, untethering them from the main domains, makes it easier to template, and the separation of the feature more clear. It is a routing feature after all. This also means users can route "any" domain, without it being configured inside the main reverse proxy.

- [x] This all needs some more testing. Looks good to me though.
- [x] https://github.com/opnsense/docs/pull/581

### CODE REVIEW

**Comment by Monviech** [comment]
@fichtner Okay, now it is /really/ done. ^^

**Comment by Monviech** [comment]
@fichtner Okay this is as good as it gets with a first version and only me as the sole tester. Please merge, I won't touch this anymore now and I want to think of other things now. xD

File: www/caddy/src/opnsense/service/templates/OPNsense/Caddy/Caddyfile [hunk 1]
```diff
@@ -234,6 +247,34 @@
...
+{#
+#   When Layer4 is active, under certain conditions
+#   this workaround is necessary so Caddy listens on the default ports.
+#   Otherwise the Layer4 Routes will not work, since there is no port listening for them.
+#}
+{% set hasReverseProxyEntry = helpers.toList('Pischem.caddy.reverseproxy.reverse') | selectattr('enabled', 'equalto', '1') | list | length > 0 %}
+
+{% if enableLayer4 == '1' %}
+    # Layer4 default HTTP port
+    {% if httpPort %}
```
Comment (Monviech) [inline www/caddy/src/opnsense/service/templates/OPNsense/Caddy/Caddyfile:259]: Note to myself, this part here still needs better logic.

File: www/caddy/pkg-descr [hunk 2]
```diff
@@ -27,15 +30,16 @@ Plugin Changelog
...
-Add: Run Caddy as "www" user and group, by enabling "Disable Superuser" in General Settings.
-Cleanup: Validations in general.volt are all appended to their form keys, instead of triggering a Bootstrap Dialog.
-Change: Description Fields are optional now, reducing the steps needed to create new entries.
-Add: Shortcuts to add new Domains and Handlers.
-Change: Subdomain Tab only appears when a wildcard domain has been configured.
+* Add: Run Caddy as "www" user and group, by enabling "Disable Superuser" in General Settings.
+* Cleanup: Validations in general.volt are all appended to their form keys, instead of triggering a Bootstrap Dialog.
+* Change: Description Fields are optional now, reducing the steps needed to create new entries.
+* Add: Shortcuts to add new Domains and Handlers.
+* Change: Subdomain Tab only appears when a wildcard domain has been configured.
```
Comment (fichtner) [inline www/caddy/pkg-descr:42]: just to nitpick for digestion purposes ordering for add/cleanup/change blocks would be beneficial

File: www/caddy/src/opnsense/mvc/app/models/OPNsense/Caddy/Caddy.xml [hunk 3]
```diff
@@ -1,13 +1,14 @@
 <model>
     <mount>//Pischem/caddy</mount>
-    <description>A GUI model for configuring a reverse proxy in the Caddy web server.</description>
-    <version>1.2.2</version>
+    <description>Model for Caddy HTTPS and Layer4 Reverse Proxy.</description>
```
Comment (fichtner) [inline www/caddy/src/opnsense/mvc/app/models/OPNsense/Caddy/Caddy.xml:3]: to be honest you can keep this as simple as "Caddy Proxy". I don't think it's used much or needs explaining here.

File: www/caddy/src/opnsense/mvc/app/views/OPNsense/Caddy/reverse_proxy.volt [hunk 4]
```diff
@@ -300,7 +319,7 @@
         <!-- Button group on the left -->
         <div>
             <button id="addDomainBtn" type="button" class="btn btn-secondary">{{ lang._('Step 1: Add Domain') }}</button>
-            <button id="addHandleBtn" type="button" class="btn btn-secondary">{{ lang._('Step 2: Add Upstream') }}</button>
+            <button id="addHandleBtn" type="button" class="btn btn-secondary">{{ lang._('Step 2: Add HTTP Handler') }}</button>
```
Comment (fichtner) [inline www/caddy/src/opnsense/mvc/app/views/OPNsense/Caddy/reverse_proxy.volt:322]: Does this require docs changes?

File: www/caddy/src/opnsense/service/templates/OPNsense/Caddy/Caddyfile [hunk 5]
```diff
@@ -234,6 +247,32 @@
...
+#   There are no regressions to set these ports up.
+#}
+
+{% if enableLayer4|default("0") == "1" %}
+    # Layer4 default HTTP port
+    {% if httpPort %}
+        :{{ httpPort }} {
+        }
+    {% else %}
+        :80 {
```
Comment (fichtner) [inline www/caddy/src/opnsense/service/templates/OPNsense/Caddy/Caddyfile:263]: can this be empty? why not force in the model?

File: www/caddy/src/opnsense/service/templates/OPNsense/Caddy/Caddyfile [hunk 6]
```diff
@@ -234,6 +247,32 @@
...
+        }
+    {% endif %}
+    # Layer4 default HTTPS port
+    {% if httpsPort %}
+        :{{ httpsPort }} {
+        }
+    {% else %}
+        :443 {
```
Comment (fichtner) [inline www/caddy/src/opnsense/service/templates/OPNsense/Caddy/Caddyfile:271]: same here

File: www/caddy/src/opnsense/service/templates/OPNsense/Caddy/Caddyfile (hunk 5 above)
Comment (fichtner) [reply www/caddy/src/opnsense/service/templates/OPNsense/Caddy/Caddyfile:263]: ignore, empty default with hint in form is alright

File: www/caddy/src/opnsense/service/templates/OPNsense/Caddy/Caddyfile (hunk 6 above)
Comment (fichtner) [reply www/caddy/src/opnsense/service/templates/OPNsense/Caddy/Caddyfile:271]: ignore, empty default with hint in form is alright

File: www/caddy/src/opnsense/mvc/app/views/OPNsense/Caddy/reverse_proxy.volt (hunk 4 above)
Comment (Monviech) [reply www/caddy/src/opnsense/mvc/app/views/OPNsense/Caddy/reverse_proxy.volt:322]: Yes it does, there is a PR in the doc section.
https://github.com/opnsense/docs/pull/581

File: www/caddy/pkg-descr (hunk 2 above)
Comment (Monviech) [reply www/caddy/pkg-descr]: Cleanup up the whole file. :D

File: www/caddy/src/opnsense/mvc/app/models/OPNsense/Caddy/Caddy.xml (hunk 3 above)
Comment (Monviech) [reply www/caddy/src/opnsense/mvc/app/models/OPNsense/Caddy/Caddy.xml]: Changed description.

**Comment by fichtner** [comment]
Merged, thanks!

**Pull request merged**